lib/model/CDataGatherer.cc

```diff
@@ -877,43 +818,39 @@ bool CDataGatherer::restoreBucketGatherer(const std::string& summaryCountFieldNa
     do {
         const std::string& name = traverser.name();
         if (name == CBucketGatherer::EVENTRATE_BUCKET_GATHERER_TAG) {
-            TBucketGathererPtr gatherer{boost::make_unique<CEventRateBucketGatherer>(
+            m_BucketGatherer = boost::make_unique<CEventRateBucketGatherer>(
                 *this, summaryCountFieldName, personFieldName, attributeFieldName,
-                valueFieldName, influenceFieldNames, traverser)};
-            if (gatherer == nullptr) {
-                LOG_ERROR(<< "Failed to create gatherer");
+                valueFieldName, influenceFieldNames, traverser);
+            if (m_BucketGatherer == nullptr) {
+                LOG_ERROR(<< "Failed to create event rate bucket gatherer");
                 return false;
             }
-            m_Gatherers.push_back(std::move(gatherer));
         } else if (name == CBucketGatherer::METRIC_BUCKET_GATHERER_TAG) {
-            TBucketGathererPtr gatherer{boost::make_unique<CMetricBucketGatherer>(
+            m_BucketGatherer = boost::make_unique<CMetricBucketGatherer>(
                 *this, summaryCountFieldName, personFieldName, attributeFieldName,
-                valueFieldName, influenceFieldNames, traverser)};
-            if (gatherer == nullptr) {
-                LOG_ERROR(<< "Failed to create gatherer");
+                valueFieldName, influenceFieldNames, traverser);
+            if (m_BucketGatherer == nullptr) {
+                LOG_ERROR(<< "Failed to create metric bucket gatherer");
                 return false;
             }
-            m_Gatherers.push_back(std::move(gatherer));
         }
     } while (traverser.next());
 
     return true;
 }
 
 void CDataGatherer::persistBucketGatherers(core::CStatePersistInserter& inserter) const {
-    for (const auto& gatherer : m_Gatherers) {
-        const std::string& tag = gatherer->persistenceTag();
-        if (tag == CBucketGatherer::EVENTRATE_BUCKET_GATHERER_TAG) {
-            const CEventRateBucketGatherer* gatherer_ =
-                dynamic_cast<const CEventRateBucketGatherer*>(gatherer.get());
-            inserter.insertLevel(tag, boost::bind(&CEventRateBucketGatherer::acceptPersistInserter,
-                                                  boost::cref(gatherer_), _1));
-        } else if (tag == CBucketGatherer::METRIC_BUCKET_GATHERER_TAG) {
-            const CMetricBucketGatherer* gatherer_ =
-                dynamic_cast<const CMetricBucketGatherer*>(gatherer.get());
-            inserter.insertLevel(tag, boost::bind(&CMetricBucketGatherer::acceptPersistInserter,
-                                                  boost::cref(gatherer_), _1));
-        }
+    const std::string& tag = m_BucketGatherer->persistenceTag();
+    if (tag == CBucketGatherer::EVENTRATE_BUCKET_GATHERER_TAG) {
```

Comment:
This probably doesn't need to cast and use separate code paths since `bind` behaves polymorphically, i.e.
I think
```
inserter.insertLevel(m_BucketGatherer->persistenceTag(),
                     boost::bind(&CBucketGatherer::acceptPersistInserter, m_BucketGatherer.get(), _1));
```
works. Persist and restore unit tests will fail if it doesn't so you can try it out safely.

Reply:
Done!